File: src/main/java/com/codethen/algorithms/exercises/LongestPalindrome.java
package com.codethen.algorithms.exercises;

/**
 * https://leetcode.com/problems/longest-palindromic-substring/
 */
public class LongestPalindrome {

    public static void main(String... args) {

        System.out.println( longestPalindrome("z234a5abbba54a32z") );
    }

    private static String longestPalindrome(String s) {

        if (s.isEmpty()) return "";

        int[] longest = {0, 1};

        for (int i = 0; i < s.length(); i++) {
            int[] even = findPalindrome(s, i, i+1);
            int[] odd = findPalindrome(s, i-1, i+1);
            if ((even[1] - even[0]) > (longest[1] - longest[0])) longest = even;
            if ((odd[1] - odd[0]) > (longest[1] - longest[0])) longest = odd;
        }

        return s.substring(longest[0], longest[1]);
    }

    private static int[] findPalindrome(String s, int i, int j) {
        while (i >= 0 && j < s.length() && s.charAt(i) == s.charAt(j)) {
            i--;
            j++;
        }
        return new int[]{i+1, j};
    }
}

File: src/main/java/com/codethen/algorithms/exercises/LongestPalindrome.kt
package com.codethen.algorithms.exercises

fun main() {

    println( longestPalindrome("z234a5abbba54a32z") )
}

fun longestPalindrome(s: String): String {

    var longest : List<Int> = listOf(0, 1)

    for (i in s.indices) {

        val odd = findPalindrome(s, i-1, i+1)
        val even = findPalindrome(s, i, i+1)

        val comp = Comparator<List<Int>> {
            x, y -> (x[1]-x[0]) - (y[1]-y[0])
        }

        longest = maxOf(longest, odd, comp)
        longest = maxOf(longest, even, comp)

        println("odd: $odd, even: $even, longest: $longest")
    }

    return s.substring(longest[0], longest[1])
}

fun findPalindrome(s: String, i: Int, j: Int): List<Int> {

    var ii = i
    var jj = j

    while (ii >= 0 && jj < s.length && s[ii] == s[jj]) {
        ii--
        jj++
    }

    return listOf(ii+1, jj)
}
